### PR TITLE
Split ForwardMode into separate classes

### PR DESCRIPTION
**Pull request by vaithak**

### CODE REVIEW

**Comment by vaithak** [comment]
will add more test cases to make codecov happy, but some of these will require changes in this PR: https://github.com/vgvassilev/clad/pull/571

**Comment by vgvassilev** [comment]
@vaithak do you think you can write some tests for the missed branches that codecov reports. They are not due to this PR but we can right a historical wrong :)

**Review by parth-07 — COMMENTED** [review]
@vaithak This looks great!

I will need a bit more time to review this pull request.

**Comment by vaithak** [comment]
> @vaithak do you think you can write some tests for the missed branches that codecov reports. They are not due to this PR but we can right a historical wrong :)

Sure, I am working on this, that's why I was able to find issues with comma expressions and the ternary operator case 😅 , these didn't had tests till now for forward mode AD.

File: lib/Differentiator/CMakeLists.txt [hunk 1]
```diff
@@ -24,7 +24,9 @@ add_llvm_library(cladDifferentiator
   ConstantFolder.cpp
   DerivativeBuilder.cpp
   DiffPlanner.cpp
+  BaseForwardModeVisitor.cpp
   ForwardModeVisitor.cpp
+  VectorForwardModeVisitor.cpp
```
Comment (parth-07) [inline lib/Differentiator/CMakeLists.txt:29]: Minor: Please sort the filenames here.

**Review by parth-07 — CHANGES_REQUESTED** [review]
It looks great overall.

File: lib/Differentiator/VectorForwardModeVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  // Add an empty return statement to the array of statements.
+  returnStmts.push_back(
+      m_Sema.ActOnReturnStmt(noLoc, nullptr, getCurrentScope()).get());
+
+  // Create a return statement from the compound statement.
+  Stmt* returnStmt = MakeCompoundStmt(returnStmts);
+  return StmtDiff(returnStmt);
+}
+
+VarDeclDiff VectorForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/VectorForwardModeVisitor.cpp:204]: the difference was in prefix assigned tot he derived variable, so created an extra overloaded function for this which takes a prefix as a parameter.

File: lib/Differentiator/CMakeLists.txt (hunk 1 above)
Comment (vaithak) [reply lib/Differentiator/CMakeLists.txt]: done 👍🏼

File: lib/Differentiator/VectorForwardModeVisitor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/VectorForwardModeVisitor.cpp:204]: Please ignore the previous comment, currently the difference is in just prefix and type of derived variable, but it will be more visible in [this PR](https://github.com/vgvassilev/clad/pull/570), where the initialisation style will also change for vector mode.
So, I think it is better to override this.

**Comment by vaithak** [comment]
added test for most of the uncovered cases to increase coverage (ignoring some failure edge cases)

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,1901 @@
...
+  } else if (incDiff.getExpr_dx() && incDiff.getExpr()) {
+    // If no declarations are required and only two Expressions are produced,
+    // join them with comma expression.
+    if (!isUnusedResult(incDiff.getExpr_dx()))
+      incResult = BuildOp(BO_Comma, BuildParens(incDiff.getExpr_dx()),
+                          BuildParens(incDiff.getExpr()));
+    else
+      incResult = incDiff.getExpr();
+  } else if (incDiff.getExpr()) {
+    incResult = incDiff.getExpr();
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:548]: Don't we want to write a test for this branch?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,1901 @@
...
+
+  StmtDiff Result =
+      (Block->size() == 1) ? StmtDiff(forStmtDiff) : StmtDiff(Block);
+  return Result;
+}
+
+StmtDiff BaseForwardModeVisitor::VisitReturnStmt(const ReturnStmt* RS) {
+  // If there is no return value, we must not attempt to differentiate
+  if (!RS->getRetValue())
+    return nullptr;
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:585]: That should be easy to test with having a `void f(double d) { return; }`?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 5]
```diff
@@ -0,0 +1,1901 @@
...
+        ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
+    return StmtDiff(clonedME, zero);
+  } else {
+    auto zero =
+        ConstantFolder::synthesizeLiteral(m_Context.DoubleTy, m_Context, 0);
+    auto baseDiff = Visit(ME->getBase());
+    // No derivative found for base. Therefore, derivative is 0.
+    if (isa<IntegerLiteral>(baseDiff.getExpr_dx()) ||
+        isa<FloatingLiteral>(baseDiff.getExpr_dx()))
+      return {clonedME, zero};
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:623]: @parth-07 do you have an idea how to test this case?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 6]
```diff
@@ -0,0 +1,1901 @@
...
+        auto result_at_i = BuildArraySubscript(it->second, clonedIndices);
+        return StmtDiff{cloned, result_at_i};
+      }
+
+      VD = decl;
+    }
+  } else if (isa<MemberExpr>(clonedBase->IgnoreParenImpCasts())) {
+    auto derivedME = cloneDiff.getExpr_dx();
+    if (!isa<MemberExpr>(derivedME->IgnoreParenImpCasts())) {
+      return {cloned, zero};
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:689]: @parth-07 and here?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 7]
```diff
@@ -0,0 +1,1901 @@
...
+    }
+    auto derivedAS = BuildArraySubscript(derivedME, clonedIndices);
+    return {cloned, derivedAS};
+  } else {
+    if (!isa<DeclRefExpr>(clonedBase->IgnoreParenImpCasts()))
+      return StmtDiff(cloned, zero);
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:695]: @parth-07 and probably here?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 8]
```diff
@@ -0,0 +1,1901 @@
...
+  // numerical diff to use correct declaration context.
+  if (forCustomDerv) {
+    DeclContext* outermostDC = utils::GetOutermostDC(m_Sema, originalFnDC);
+    // FIXME: We should ideally construct nested name specifier from the
+    // found custom derivative function. Current way will compute incorrect
+    // nested name specifier in some cases.
+    if (outermostDC &&
+        outermostDC->getPrimaryContext() == NSD->getPrimaryContext()) {
+      utils::BuildNNS(m_Sema, originalFnDC, SS);
+      DC = originalFnDC;
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:873]: @grimmmyshini do you have an idea how to test this branch?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 9]
```diff
@@ -0,0 +1,1901 @@
...
+    return StmtDiff(op, BuildOp(opKind, diff.getExpr_dx()));
+  } else {
+    // This only adds support for differentiating the following expression:
+    // ```
+    // *this;
+    // ```
+    if (opKind == UnaryOperatorKind::UO_Deref) {
+      if (isa<CXXThisExpr>(UnOp->getSubExpr()->IgnoreParenImpCasts())) {
+        Expr* derivedE = BuildOp(opKind, diff.getExpr_dx());
+        return {op, derivedE};
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1136]: @parth-07 and here?

File: lib/Differentiator/BaseForwardModeVisitor.cpp [hunk 10]
```diff
@@ -0,0 +1,1901 @@
...
+  return nullptr;
+}
+
+static void setSwitchCaseSubStmt(SwitchCase* SC, Stmt* subStmt) {
+  if (auto caseStmt = dyn_cast<CaseStmt>(SC)) {
+    caseStmt->setSubStmt(subStmt);
+  } else if (auto defaultStmt = dyn_cast<DefaultStmt>(SC)) {
+    defaultStmt->setSubStmt(subStmt);
+  } else {
+    assert(0 && "Unsupported switch case statement");
```
Comment (vgvassilev) [inline lib/Differentiator/BaseForwardModeVisitor.cpp:1522]: Is that reachable at all?

**Comment by vgvassilev** [comment]
@vaithak did you run clang-format on this patch?

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 7 above)
Comment (parth-07) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:695]: This example reaches this code:

```cpp
#include "clad/Differentiator/Differentiator.h"

struct Array {
    double data[5];
};

Array array;

double fn(double i, double j) {
    return array.data[0] * i + array.data[1] * j;
}

int main() {
    auto d_fn = clad::differentiate(fn, "i");
}
```

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 9 above)
Comment (parth-07) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1136]: It seems that this `if` block can be safely removed. We later added proper support for dereference operator, which made this special case redundant.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 10 above)
Comment (parth-07) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1522]: It's an assert. Hopefully, it's not reachable :).

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 7 above)
Comment (parth-07) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:695]: I cannot think of a case for this right now. But it might be possible in some obscure cases.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:548]: I couldn't find any expression such that the original expression is not null, but it's derivative expression is null.
Can this branch be removed ? cc: @parth-07

**Comment by vaithak** [comment]
> @vaithak did you run clang-format on this patch?

yup 👍🏼

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 5 above)
Comment (vaithak) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:623]: Thanks for the example, added this as a test.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 9 above)
Comment (vaithak) [reply lib/Differentiator/BaseForwardModeVisitor.cpp]: removed this condition

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 3 above)
Comment (parth-07) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:548]: No, this cannot be removed. This example reaches this branch:

```cpp
#include "clad/Differentiator/Differentiator.h"

void increment(int &i) {
    ++i;
}

double fn(double i, double j) {
    double res = 0;
    for (int i=0; i < 5; increment(i))
        res += i*j;
    return res;
}

int main() {
    auto d_fn = clad::differentiate(fn, "i");
}
```

File: lib/Differentiator/VectorForwardModeVisitor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (parth-07) [reply lib/Differentiator/VectorForwardModeVisitor.cpp:204]: Okay, got it.

**Review by parth-07 — APPROVED** [review]
Looks good overall. Please add the remaining test cases if any and squash the commits into one.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 10 above)
Comment (vgvassilev) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1512]: Perhaps also issue an unsupported diagnostics?

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:548]: Oh, nice! I wrongly tried using return directly as it has a derivative expression as null.
Thanks for this 👍🏼 , added.

File: lib/Differentiator/BaseForwardModeVisitor.cpp (hunk 10 above)
Comment (vaithak) [reply lib/Differentiator/BaseForwardModeVisitor.cpp:1512]: This is inside a static function, so `diag` can't be used here.

**Review by parth-07 — APPROVED** [review]
Thank you for the great work. LGTM.

**Review by vgvassilev — APPROVED** [review]
Great work!
